### PR TITLE
Request handler refactor

### PR DESCRIPTION
**Pull request by apfong**
Covers a lot of changes to get RequestHandling working with the new common API (Request/Response Objects, config format, our new dispatching mechanism through RequestRouter, and [handler registration](https://github.com/jfarrell468/registerer)).

Also did a little bit of cleanup on logging.

### CODE REVIEW

File: server_config.h [hunk 1]
```diff
@@ -17,16 +17,17 @@ struct container_hash {
 
 class ServerConfig {
   public:
+    ServerConfig();
     ServerConfig(NginxConfig config);
     bool Build();
```
Comment (KyCodeHuynh) [inline server_config.h]: Minor, but this should be `build()` to match our conventions.

File: request_handlers.cc [hunk 2]
```diff
@@ -12,88 +15,78 @@
...
-  response_buffer = new char[response_buffer_size + 1];
-  // std::memset((void *)&response_buffer, 0, sizeof(response_buffer));
-
-  // Copy in headers, the original request, and a terminating nullbyte
-  response_header.copy(response_buffer, header_size);
-  std::memcpy(&response_buffer[header_size], request_buffer, request_buffer_size);
-  std::memcpy(&response_buffer[header_size + request_buffer_size], "\0", 1);
-
-  std::cout << "~~~~~~~~~~Response~~~~~~~~~~\n" << response_buffer << std::endl;
+  // TODO: for some reason, browser is interpreting the response as a binary file
```
Comment (KyCodeHuynh) [inline request_handlers.cc]: It's possible the `text/plain` Content-Type is not being set properly? We can verify via curl/HTTPie later that the headers are being produced correctly.

**Comment by KyCodeHuynh** [comment]
Builds should fail because we don't yet have ServerStats (which is referenced in ServerConfig tests), but why is this [linker error](https://travis-ci.org/UCLA-CS130/Mr.-Robot-et-al./builds/204805757) occurring?

**Comment by KyCodeHuynh** [comment]
Can we remove this [WIP commit](5bf4f0518e24dc685d39844472769b80586c3ba9)?

File: request_router.cc [hunk 3]
```diff
@@ -44,10 +53,18 @@ std::unique_ptr<RequestHandler> RequestRouter::routeRequest(const std::string& f
     }
     else {
       // Slice string from 0 to pos; substr takes start and length
-      std::string possible_prefix = full_uri.substr(0, pos);
-      std::cerr << "RequestRouter::routeRequest prefix substring" << possible_prefix << std::endl;
+      // (unless there is only one '/', then use the whole uri)
```
Comment (KyCodeHuynh) [inline request_router.cc]: Very nice edge-case consideration!

File: request_handlers.cc (hunk 2 above)
Comment (apfong) [inline request_handlers.cc]: Ah had actually fixed that, forgot to remove the comment. When i was changing some things, the buffer size was off by one

**Comment by apfong** [comment]
To fix the [linker error](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/55#issuecomment-282190470), we had to remove an unnecessary include of `request_handlers.h` in the `server_config_tests.cc`.

**Pull request merged**